File: packages/react-native/temp/package/ReactAndroid/src/main/java/com/facebook/react/runtime/BridgelessAtomicRef.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.runtime;

import static com.facebook.infer.annotation.Assertions.assertNotNull;

import android.annotation.SuppressLint;
import androidx.annotation.Nullable;
import com.facebook.infer.annotation.Nullsafe;
import java.util.Objects;

@Nullsafe(Nullsafe.Mode.LOCAL)
class BridgelessAtomicRef<T> {

  interface Provider<T> {
    T get();
  }

  @Nullable volatile T mValue;
  @Nullable T mInitialValue;

  enum State {
    Init,
    Creating,
    Success,
    Failure
  }

  private volatile State state;
  private volatile String failureMessage;

  public BridgelessAtomicRef(@Nullable T initialValue) {
    mValue = initialValue;
    mInitialValue = initialValue;
    state = State.Init;
    failureMessage = "";
  }

  public BridgelessAtomicRef() {
    this(null);
  }

  @SuppressLint("CatchGeneralException")
  public T getOrCreate(BridgelessAtomicRef.Provider<T> provider) {
    boolean shouldCreate = false;
    synchronized (this) {
      if (state == State.Success) {
        return get();
      }

      if (state == State.Failure) {
        throw new RuntimeException(
            "BridgelessAtomicRef: Failed to create object. Reason: " + failureMessage);
      }

      if (state != State.Creating) {
        state = State.Creating;
        shouldCreate = true;
      }
    }

    if (shouldCreate) {
      try {
        // Call provider with lock on `this` released to mitigate deadlock hazard
        mValue = provider.get();

        synchronized (this) {
          state = State.Success;
          notifyAll();
          return get();
        }
      } catch (RuntimeException ex) {
        synchronized (this) {
          state = State.Failure;
          String message = ex.getMessage();
          failureMessage = Objects.toString(message, "null");
          notifyAll();
        }

        throw new RuntimeException("BridgelessAtomicRef: Failed to create object.", ex);
      }
    }

    synchronized (this) {
      boolean wasInterrupted = false;
      while (state == State.Creating) {
        try {
          wait();
        } catch (InterruptedException ex) {
          wasInterrupted = true;
        }
      }

      if (wasInterrupted) {
        Thread.currentThread().interrupt();
      }

      if (state == State.Failure) {
        throw new RuntimeException(
            "BridgelessAtomicRef: Failed to create object. Reason: " + failureMessage);
      }

      return get();
    }
  }

  public synchronized T getAndReset() {
    T value = get();
    reset();
    return value;
  }

  public synchronized void reset() {
    mValue = mInitialValue;
    state = State.Init;
    failureMessage = "";
  }

  public synchronized T get() {
    return assertNotNull(mValue);
  }

  public synchronized @Nullable T getNullable() {
    return mValue;
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/BridgelessAtomicRef.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.runtime

import android.annotation.SuppressLint
import kotlin.concurrent.Volatile

internal class BridgelessAtomicRef<T>(
    @field:Volatile @get:Synchronized @get:JvmName("getNullable") var value: T? = null
) {

  internal fun interface Provider<T> {
    fun get(): T
  }

  internal enum class State {
    Init,
    Creating,
    Success,
    Failure
  }

  var initialValue: T? = value

  @Volatile private var state: State = State.Init

  @Volatile private var failureMessage: String = ""

  @SuppressLint("CatchGeneralException")
  @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
  fun getOrCreate(provider: Provider<T>): T {
    var shouldCreate = false
    synchronized(this) {
      if (state == State.Success) {
        return get()
      }
      if (state == State.Failure) {
        throw RuntimeException(
            "BridgelessAtomicRef: Failed to create object. Reason: $failureMessage")
      }
      if (state != State.Creating) {
        state = State.Creating
        shouldCreate = true
      }
    }

    if (shouldCreate) {
      try {
        // Call provider with lock on `this` released to mitigate deadlock hazard
        value = provider.get()

        synchronized(this) {
          state = State.Success
          (this as Object).notifyAll()
          return get()
        }
      } catch (ex: RuntimeException) {
        synchronized(this) {
          state = State.Failure
          val message = ex.message
          failureMessage = message.toString()
          (this as Object).notifyAll()
        }

        throw RuntimeException("BridgelessAtomicRef: Failed to create object.", ex)
      }
    }

    synchronized(this) {
      var wasInterrupted = false
      while (state == State.Creating) {
        try {
          (this as Object).wait()
        } catch (ex: InterruptedException) {
          wasInterrupted = true
        }
      }

      if (wasInterrupted) {
        Thread.currentThread().interrupt()
      }

      if (state == State.Failure) {
        throw RuntimeException(
            "BridgelessAtomicRef: Failed to create object. Reason: $failureMessage")
      }
      return get()
    }
  }

  @get:Synchronized
  val andReset: T
    get() {
      val value = get()
      reset()
      return value
    }

  @Synchronized
  fun reset() {
    value = initialValue
    state = State.Init
    failureMessage = ""
  }

  @Synchronized fun get(): T = checkNotNull(value)
}
